### DISCUSSION TITLE
Query Builder Files

### DISCUSSION THREAD

**Discussion by Heidinger35** [DaxStudio/DaxStudio#1115] [Q&A] [answered]
Hello. My org heavily uses query builder within DAX Studio. Teams are hoping to save the DAX Studio files that maintain the columns, measures, and filters users to build the query so adjustments can be made without having to rebuild the query each time. On some occasions when a file is saved, two files are generated:

1. DAX Query file (.dax)
2. QUERYBUILDER file (.queryBuilder)

This second file seems to be needed in order for the .dax file to maintain columns, measures, and filters that were used to build the query. Can someone please provide insight on:

- Why are the .queryBuilder files generated sometimes and not others?
- Is there a setting to always have a .queryBuilder file generated?
- Any additional thoughts / best practices on saving a DAX Studio instance that maintains query builder selections?

Any help is greatly appreciated!
![DAX Query Builder Files](https://github.com/DaxStudio/DaxStudio/assets/56665555/0117fd83-ce0e-44ee-b2c4-b0b039633525)

**Comment by dgosbell** [comment]
> Why are the .queryBuilder files generated sometimes and not others?

the .queryBuilder files should only be generated if the query builder is open. If you close the query builder this file should not be generated

> Is there a setting to always have a .queryBuilder file generated?

No, as above if the query builder is open it should save, if it's not open it should not save.

> Any additional thoughts / best practices on saving a DAX Studio instance that maintains query builder selections?

There is a new file format you can choose when doing a "save as" which is a .daxx file this will save both the .dax and .querybuilder file internally in a zip format so that might be an option if you don't want the two files saved all the time.

**Reply by Heidinger35** [reply]
Hi Darren,

I am not experiencing the behavior you have outlined. While saving, I always have the query builder open and am not getting the querybuilder files to generate. I only get a .dax file.

I have tried saving as both "DAX documents" and "DAXX documents" multiple times and both only generate a .dax file. When this file is opened, the "Columns / Measures" and "Filters" areas are empty and I have to rebuild thew query from scratch. 

After I select save, the log gives me the error in the attached image. Does this have anything to do with it?

![image](https://github.com/DaxStudio/DaxStudio/assets/56665555/fc9c5204-f5bb-4b40-a0f8-d95c0a7dc92c)

**Update - If there is not a measure included in the query, the saved file will open with query builder fields populated. It is only when a measure is included that this error is generated and the query builder opens empty in the saved file. Also, there is still only one file being generated from the DAXX save option but again, this does will open with query builder populated as long as a measure is not included in the query. 

Insight on why this is happening / how I can have query builder open populated when a measure is used would be helpful.

**Comment by dgosbell** [comment]
So the saving error is definitely what is stopping the .querybuilder file from being created. So this is a DAX Studio bug of some sort.

What sort of data source are you connecting to? If you hover over the database can you check if you have admin rights?
![image](https://github.com/DaxStudio/DaxStudio/assets/8241830/4ce69174-1533-4a7b-b9f0-f2b77b1103ce)

I just deployed a dataset to the service then logged into DAX Studio with a secondary user that only had View & Build rights and I was able to reproduce this issue, so you this might be a permissions related. Originally readonly users used to be able to see measure expressions in the metadata, but that was blocked a year or more ago and this looks to be a bug linked to that change which would not be hard to fix.

**Reply by Heidinger35** [reply]
The data source is a Power BI dataset that uses direct query against Snowflake. I do **not** have admin rights. 

Should I use the "Report a Bug" feature to report this? How often are updates made / released?

**Reply by dgosbell** [reply, marked as answer]
Sorry I missed looping back on this. This should be fixed in the current release.
And if you do see an error like this opening a file it is a good idea to use the Report a Bug feature

**Reply by Heidinger35** [reply]
Yep this was fixed and is working great!
